MalomAPI_cpp/perfect_wrappers.h

```diff
@@ -0,0 +1,362 @@
+/*
+Malom, a Nine Men's Morris (and variants) player and solver program.
+Copyright(C) 2007-2016  Gabor E. Gevay, Gabor Danner
+Copyright (C) 2023 The Sanmill developers (see AUTHORS file)
+
+See our webpage (and the paper linked from there):
+http://compalg.inf.elte.hu/~ggevay/mills/index.php
+
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef PERFECT_WRAPPER_H_INCLUDED
+#define PERFECT_WRAPPER_H_INCLUDED
+
+#include "perfect_common.h"
+#include "perfect_hash.h"
+#include "perfect_symmetries.h"
+#include "perfect_debug.h"
+#include "perfect_sector.h"
+#include "perfect_sector_graph.h"
+
+#include <cassert>
+#include <cmath> // for factorial function
+#include <iostream>
+#include <map>
+#include <set>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+namespace Wrappers {
+
+class WSector;
+
+extern std::unordered_map<id, int> sector_sizes;
+static int f_inv_count[] {1,     4,     30,    158,    757,    2830,  8774,
+                          22188, 46879, 82880, 124124, 157668, 170854};
+
+struct WID
+{
+    int W, B, WF, BF;
+    WID(int W, int B, int WF, int BF)
+        : W(W)
+        , B(B)
+        , WF(WF)
+        , BF(BF)
+    { }
+    WID(id id)
+        : W(id.W)
+        , B(id.B)
+        , WF(id.WF)
+        , BF(id.BF)
+    { }
+    ::id tonat() { return ::id(W, B, WF, BF); }
+    void negate();
+    WID operator-(WID s);
+
+    std::string ToString() { return this->tonat().to_string(); }
+
+    int GetHashCode() { return (W << 0) | (B << 4) | (WF << 8) | (BF << 12); }
+
+private:
+    static int64_t factorial(int n)
```

Comment:
Unfortunately, `int64_t` won't be large enough for this. `nCr(n,r)` calls `factorial(n)`, and `size()` calls `nCr(24 - W, B)`, where `W` can be as small as 3. So the factorial function needs to be able to compute `factorial(21)`, which is `51090942171709440000`, which doesn't fit in `int64_t`, or even `uint64_t`, since `2^64 = 18446744073709551616`.

You might be able to simply delete these functions (`factorial`, `nCr`, `size`) from the C++ code (and the `sector_sizes` variable). If I remember correctly, `size()` is used only by the `Controller` project, but not the GUI or the API. (I can't easily verify this right now, because I don't have that computer with me where I have my development environment installed.)

Reply:
We've confirmed that the mentioned functions and variable are not used in our project. They will be removed in the next update.